src/glang/driver.py

```diff
@@ -77,115 +82,229 @@ def __init__(self):
+    ) -> None:
+        # Stop parsing and exit after printing the host's target. If we keep
+        # parsing, then the parser complains that `input`, a required position
+        # argument, is missing.
+        print(get_host_target())
+        parser.exit()
 
-    will_emit_llvm = args.emit_llvm or args.emit_everything
-    will_emit_llvm_to_stdout = args.emit_llvm_to_stdout
-    will_emit_optimized_llvm = args.emit_optimized_llvm or args.emit_everything
-    will_emit_binary = (
-        not (args.emit_llvm or args.emit_llvm_to_stdout) or args.emit_everything
-    )
 
-    llvm_output = args.input.with_suffix(".ll")
-    optimized_llvm_output = llvm_output.with_suffix(".opt.ll")
-    binary_output = Path("a.out")
+@dataclass(frozen=True)
+class Stage:
+    source: Path | bytes
+    derived_from: "Stage | None"
 
-    load_target_config(args.target)
+    def get_bytes(self) -> bytes:
+        if isinstance(self.source, bytes):
+            return self.source
 
-    parent_dir = Path(__file__).parent.resolve()
-    lib_dir = parent_dir / "lib"
-    target_dir = lib_dir / "std" / get_target()
+        with open(self.source, "rb") as f:
+            return f.read()
 
-    if not args.nostdlib:
-        args.include_path.append(lib_dir)
-        args.include_path.append(target_dir)
+    def get_file(self) -> Path:
+        if isinstance(self.source, Path):
+            return self.source
 
-    # -o defaults to binary output path
-    if args.output:
-        if will_emit_binary:
-            binary_output = args.output
-        else:
-            llvm_output = args.output
+        path = Path(mkstemp(dir=Path(global_tmp_dir.name))[1])
+        path.write_bytes(self.source)
+        return path
 
-    # Compile to ir
-    ir = generate_ir_from_source(
-        args.input, args.include_path, debug_compiler=args.debug_compiler
-    )
+    def get_top_level_source_file(self) -> Path:
+        if self.derived_from is None:
+            return self.get_file()
+        return self.derived_from.get_top_level_source_file()
 
-    if will_emit_llvm:
-        with llvm_output.open("w", encoding="utf-8") as file:
-            file.write(ir)
 
-    if will_emit_llvm_to_stdout:
-        sys.stdout.write(ir)
+class ELF_Binary(Stage):
+    pass
 
-    # Use clang to finish compile
-    assert args.optimize != "2", "@DaveDuck321's coward assert!"
 
-    clang = getenv("GRAPHENE_CLANG_CMD", "clang")
+class Assembly(Stage):
+    def compile(self, args: DriverArguments) -> ELF_Binary:
+        expanded = run_checked(
+            [
+                getenv("GRAPHENE_CLANG_CMD", "clang"),
+                "-E",
+                "-xassembler-with-cpp",
+                "-",
+            ],
+            stdin=self.get_bytes(),
+        )
+        compiled = run_checked(
+            [
+                getenv("GRAPHENE_LLVM_MC_CMD", "llvm-mc"),
+                "--filetype=obj",
+                f"--triple={get_target_triple()}",
```

Comment:
It's not a huge deal but it seems like the source location generated by this is `stdin` rather than `runtime.S` due to the pipe from the macro expansion :cry:

Reply:
Ugh, that kinda sucks. If there's no way around, should we just use `clang` here?

Reply:
Oh, nvm I was just bad. It was just picking up Clang's annotations in the macro expansion. Fixing the debug was as simple as just giving `clang -E` a path rather than a pipe.

Reply:
oh duh, yeah that makes sense